File: examples/python/comm-ready/comm-ready.py
# Copyright (c) 2010-2022. The SimGrid Team. All rights reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the license (GNU LGPL) which comes with this package.

from argparse import ArgumentParser
from typing import List
import sys

from simgrid import Actor, Comm, Engine, Mailbox, this_actor


FINALIZE_MESSAGE = "finalize"


def create_parser() -> ArgumentParser:
    parser = ArgumentParser()
    parser.add_argument(
        '--platform',
        type=str,
        required=True,
        help='path to the platform description'
    )
    return parser


def get_peer_mailbox(peer_id: int) -> Mailbox:
    return Mailbox.by_name(f"peer-{peer_id}")


def peer(my_id: int, message_count: int, payload_size: int, peers_count: int):
    my_mailbox: Mailbox = get_peer_mailbox(my_id)
    my_mailbox.set_receiver(Actor.self())
    pending_comms: List[Comm] = []
    # Start dispatching all messages to peers others that myself
    for i in range(message_count):
        for peer_id in range(peers_count):
            if peer_id != my_id:
                peer_mailbox = get_peer_mailbox(peer_id)
                message = f"Message {i} from peer {my_id}"
                this_actor.info(f"Send '{message}' to '{peer_mailbox.name}'")
                pending_comms.append(peer_mailbox.put_async(message, payload_size))

    # Start sending messages to let peers know that they should stop
    for peer_id in range(peers_count):
        if peer_id != my_id:
            peer_mailbox = get_peer_mailbox(peer_id)
            payload = str(FINALIZE_MESSAGE)
            pending_comms.append(peer_mailbox.put_async(payload, payload_size))
            this_actor.info(f"Send '{payload}' to '{peer_mailbox.name}'")
    this_actor.info("Done dispatching all messages")

    # Retrieve all the messages other peers have been sending to me until I receive all the corresponding "Finalize"
    # messages
    pending_finalize_messages = peers_count - 1
    while pending_finalize_messages > 0:
        if my_mailbox.ready:
            start = Engine.clock
            received: str = my_mailbox.get()
            waiting_time = Engine.clock - start
            if waiting_time != 0.0:
                raise AssertionError(f"Expecting the waiting time to be 0.0 because the communication was supposedly "
                                     f"ready, but got {waiting_time} instead")
            this_actor.info(f"I got a '{received}'.")
            if received == FINALIZE_MESSAGE:
                pending_finalize_messages -= 1
        else:
            this_actor.info("Nothing ready to consume yet, I better sleep for a while")
            this_actor.sleep_for(0.01)

    this_actor.info("I'm done, just waiting for my peers to receive the messages before exiting")
    Comm.wait_all(pending_comms)
    this_actor.info("Goodbye now!")


def main():
    settings = create_parser().parse_known_args()[0]
    e = Engine(sys.argv)
    e.load_platform(settings.platform)
    Actor.create("peer", e.host_by_name("Tremblay"), peer, 0, 2, int(5e7), 3)
    Actor.create("peer", e.host_by_name("Ruby"), peer, 1, 6, int(2.5e5), 3)
    Actor.create("peer", e.host_by_name("Perl"), peer, 2, 0, int(5e7), 3)
    e.run()


if __name__ == "__main__":
    main()

File: examples/java/comm_ready/comm_ready.java
/* Copyright (c) 2010-2025. The SimGrid Team. All rights reserved.          */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

/* This example shows how to use Mailbox.ready() to check for completed communications.
 *
 * We have a number of peers which send and receive messages in two phases:
 * -> sending phase:   each one of them sends a number of messages to the others followed
 *                     by a single "finalize" message.
 * -> receiving phase: each one of them receives all the available messages that reached
 *                     their corresponding mailbox until it has all the needed "finalize"
 *                     messages to know that no more work needs to be done.
 *
 * To avoid doing an await() over the ongoing communications, each peer makes use of the
 * Mailbox.ready() method. If it returns true then a following get() will fetch the
 * message immediately, if not the peer will sleep for a fixed amount of time before checking again.
 *
 */
import org.simgrid.s4u.*;

class peer extends Actor {
  int myId;
  int messagesCount;
  int payloadSize;
  int peersCount;
  peer(int myId_, int messagesCount_, double payloadSize_, int peersCount_)
  {
    myId          = myId_;
    messagesCount = messagesCount_;
    payloadSize   = (int)payloadSize_;
    peersCount    = peersCount_;
  }
  public void run() throws SimgridException
  {
    Engine e = this.get_engine();
    /* Set myself as the persistent receiver of my mailbox so that messages start flowing to me as soon as they are put
     * into it */
    Mailbox myMbox = e.mailbox_by_name("peer-" + myId);
    myMbox.set_receiver(Actor.self());

    ActivitySet pending_comms = new ActivitySet();

    /* Start dispatching all messages to peers others that myself */
    for (int i = 0; i < messagesCount; i++) {
      for (int peer_id = 0; peer_id < peersCount; peer_id++) {
        if (peer_id != myId) {
          Mailbox mbox   = e.mailbox_by_name("peer-" + peer_id);
          String payload = "Message " + i + " from peer " + myId;
          Engine.info("Send '" + payload + "' to '" + mbox.get_name() + "'");
          /* Create a communication representing the ongoing communication */
          pending_comms.push(mbox.put_async(payload, payloadSize));
        }
      }
    }

    /* Start sending messages to let peers know that they should stop */
    for (int peer_id = 0; peer_id < peersCount; peer_id++) {
      if (peer_id != myId) {
        Mailbox mbox   = e.mailbox_by_name("peer-" + peer_id);
        String payload = "finalize";
        pending_comms.push(mbox.put_async(payload, payloadSize));
        Engine.info("Send 'finalize' to 'peer-" + peer_id + "'");
      }
    }
    Engine.info("Done dispatching all messages");

    /* Retrieve all the messages other peers have been sending to me until I receive all the corresponding "Finalize"
     * messages */
    long pendingFinalizeMessages = peersCount - 1;
    while (pendingFinalizeMessages > 0) {
      if (myMbox.ready()) {
        double start        = Engine.get_clock();
        String received     = (String)myMbox.get();
        double waitingTime  = Engine.get_clock() - start;
        if (waitingTime > 0)
          Engine.die("Expecting the waiting time to be 0 because the communication was supposedly ready, but got " +
                     waitingTime + " instead");
        Engine.info("I got a '" + received + "'.");

        if (received.equals("finalize"))
          pendingFinalizeMessages--;

      } else {
        Engine.info("Nothing ready to consume yet, I better sleep for a while");
        sleep_for(.01);
      }
    }

    Engine.info("I'm done, just waiting for my peers to receive the messages before exiting");
    pending_comms.await_all();

    Engine.info("Goodbye now!");
  }
}

class comm_ready {
  public static void main(String[] args)
  {
    Engine e = new Engine(args);
    e.load_platform(args[0]);
    e.host_by_name("Tremblay").add_actor("peer", new peer(0, 2, 5e7, 3));
    e.host_by_name("Ruby").add_actor("peer", new peer(1, 6, 2.5e5, 3));
    e.host_by_name("Perl").add_actor("peer", new peer(2, 0, 5e7, 3));

    e.run();

    // The following call is useless in your code, but our continuous integration uses it to track memleaks
    e.force_garbage_collection();
  }
}